### ISSUE TITLE
:terminal should not use &shell

### ISSUE THREAD

**Issue by splinterofchaos** [neovim/neovim#2292]
I use fish as a default shell, but sh as my shell for vim scripts because many of the plugins I use assume the standard sh syntax. fish does work, actually very well, with `:terminal`.

According to [one SO answer](http://stackoverflow.com/questions/11059067/what-is-the-nix-command-to-view-a-users-default-login-shell), the login shell can be obtained by running

``` bash
$ getent passwd $LOGNAME | cut -d: -f7
```

...and for Mac, [another SO answer](http://stackoverflow.com/questions/16375519/how-to-get-default-shell) recommended the same, but using this:

``` bash
$ grep ^$(id -un): /etc/passwd | cut -d : -f 7-
```

although the `^$(id -un)` syntax isn't fish-compatible and `grep` may output more than one username.

**Comment by HarmtH** [comment]
That's nice for scripts, but there are better ways for C programs. For example, [here](https://github.com/linuxmint/gnome-apps/blob/master/gnome-terminal/src/eggshell.c) you can see how gnome-terminal determines the user's default shell.

**Comment by ZyX-I** [comment]
I do not agree. Like Vim NeoVim uses `SHELL` environment variable for `&shell` which is [defined](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html) as

> This variable shall represent a pathname of the user's *preferred command language interpreter*. If this interpreter does not conform to the Shell Command Language in the Shell and Utilities volume of IEEE Std 1003.1-2001, Chapter 2, Shell Command Language, utilities may behave differently from those described in IEEE Std 1003.1-2001.

. `fish` does not alter the value of `$SHELL` and it is also explicitly mentioned that interpreter is expected to be POSIX-conformant.

Login shell is completely irrelevant and may even be set to `/sbin/nologin` so that only things like `sudo` or `ssh` with explicit command may be used to run something. If NeoVim was using entry from /etc/passwd it would not be able to run shell commands under some not uncommon circumstances.

**Comment by ZyX-I** [comment]
I am wrong about `/sbin/nologin`: it disables `ssh`.

**Comment by ZyX-I** [comment]
But this does not disable `sudo`.

**Comment by splinterofchaos** [comment]
@HarmtH said:
> here you can see how gnome-terminal determines the user's default shell.

that seems reasonable, and it would solve my issue.

@ZyX-I 
> fish does not alter the value of $SHELL 

It does for me, and I checked my configuration--I don't manually set it at all.

> and it is also explicitly mentioned that interpreter is expected to be POSIX-conformant.

True, but that begs the question: why not use bash or zsh? Plain ol' sh is fast, small, and I never need all the fancier features for vim plugins. Even if nvim used `$SHELL` instead, and I manually set `$SHELL` to zsh for compliance, that would be an improvement since I don't want zsh being spawned all the time.

My point is just that nvim's `shell` and the shell spawned by `:terminal` should be allowed to differ.

**Comment by ZyX-I** [comment]
Diffing makes more sense, I thought you want `&shell` to be login shell.

> It does for me, and I checked my configuration--I don't manually set it at all.

fish-2.1.1 (Gentoo amd64) and 2.1.1-1200-g793784c do not do this, config is just

```Fish
#fish_vi_mode
set fish_function_path "$HOME/.vam/powerline/powerline/bindings/fish" $fish_function_path
powerline-setup

# vim: ft=fish
```

Note that fish has such annoying thing as universal variable. If you ever made `SHELL` universal and set it to `/bin/fish` it will continue to be set to `/bin/fish` in new shells up until you set it to something else. This is “super useful” for testing (which is the main reason I have fish installed in first place) and commonly led to an unexpected behaviour. I am not much familiar with fish to remember how exactly you need to set `SHELL` to make the result exported if `SHELL` is already a universal variable.

**Comment by ZyX-I** [comment]
I would though say that `:terminal`-specific non-`&shell` value should only be used for empty terminal commands since `:terminal` is suggested to be used in scripts like fugitive.

**Comment by justinmk** [comment]
Would using `$SHELL` be good enough? Or are you asking for another set of `'shell'`, `'shellcmdflag'`, etc. options?

**Comment by splinterofchaos** [comment]
Yeah, I said login shell referencing the discussion on gitter. `$SHELL` would be good for me.

> I would though say that :terminal-specific non-&shell value should only be used for empty terminal commands since :terminal is suggested to be used in scripts like fugitive

Might some users find that confusing? For example, simple commands like `ls **` behave differently in bash than fish, and not at all in `sh`. With fish, prints the relative path of everything in the path's subdirectories, while bash prints directories with a list of files under each one. I think it's reasonable to expect `:terminal | cmd` to behave exactly like `:terminal cmd`.

I think `ls` is a good use-case too because when it detects that stdout is not a PTY, it doesn't produce an easy-to-read table.

**Comment by tarruda** [comment]
Another thing we need to do is start the shell with the login flag which seems to be the default behavior in tmux and also avoids problems like scrambled `PATH` variable as reported in gitter.

Do all shells use the -l flag to specify that it should start as a login shell?

**Comment by ZyX-I** [comment]
@tarruda AFAIK most common and forced convention is prepending a HYPHENMINUS to the shell name: e.g. `su` runs shell as `-su`. This is also how sudo works:

```
2243  execve("/usr/bin/sudo", ["sudo", "--login", "--user", "zyx"], [/* 70 vars */]) = 0
2244  execve("/bin/zsh", ["-zsh"], [/* 16 vars */]) = 0
…
```

(note: no `-l` or `--login`).

**Comment by ZyX-I** [comment]
> Might some users find that confusing? For example, simple commands like ls ** behave differently in bash than fish, and not at all in sh. With fish, prints the relative path of everything in the path's subdirectories, while bash prints directories with a list of files under each one. I think it's reasonable to expect :terminal | cmd to behave exactly like :terminal cmd.

No, it is not. `:exe 'terminal' | call feedkeys("cmd\n")` (`:terminal | cmd` will not do what I guess you meant at all) will run shell in interactive mode, `terminal cmd` will run something like `shell, "-c", cmd`. It is *already* a big difference because aliases are not used in this case.

The main reason I am suggesting what I am suggesting is that scripts are expected to run `:terminal` now under some circumstances. This means they will use `execute 'terminal command' shellescape(arg1, 1) shellescape(arg2, 1) …` and `shellescape()` escapes its value for `&shell`. It is also possible that scripts will use command substitution which differs in different shells, but I think this is much unlikely.

Or, instead of this `termescape()` function is needed which is like `shellescape(, 1)`, but for shell used by `:terminal`.

**Comment by Grimy** [comment]
I’m in favor of using `$SHELL` rather than `&shell` as the default `:terminal` command, for the same reasons as @splinterofchaos. Since &shell defaults to $SHELL anyway, the behavior would be identical for most users.

**Comment by HarmtH** [comment]
I think to have a seperate variable which also defaults to $SHELL would be nice, when :term is run without arguments.

On the note of login shells, I dont think Neovim should start a login shell, at least not by default. Login shells are meant to be run after, as the name suggests, a login. For example after 'ssh' or 'login'. At least in gnome-terminal it is an option to run as a login shell, which is off by default.

**Comment by splinterofchaos** [comment]
> The main reason I am suggesting what I am suggesting is that scripts are expected to run :terminal now under some circumstances. This means they will use execute 'terminal command' shellescape(arg1, 1) shellescape(arg2, 1) … and shellescape() escapes its value for &shell.

That's a good point... Script writers will assume standard sh syntax, which `$SHELL` may not be. Still, to get the benefits of a color terminal and PTY, which the old uses of `&shell` never needed, we could either recommend that `&shell` be a full-featured color shell or add some new config points.

We could make a `&tshell` or `&terminalshell` config and also a `tshellescape()` function or add another extra argument to `shellescape()`, allowing command shells, terminal shells, and `$SHELL` to differ independently. I'm not sure at this point whether that would introduce too much complexity, and more commands might require adaptation as well.

Finally, we'd then have the shell spawned by `:terminal` with no command, which could use `$SHELL`.

It rubs me the wrong way that at any point, we might have three different shells launched in three different ways and for different reasons, but I'm not sure this would actually cause any problems. Thoughts?

**Comment by bfredl** [comment]
For a script launching a specific command, rather than relying on `shellescape()` wouldn't it be simpler to just launch the process using the argv-list like with `jobstart` ? (at least when globbing, shell substitution etc is not required)

**Comment by wrigleyster** [comment]
What are these scripts that everyone keeps talking about? surely shell scripts should include their preferred interpreter in the hashbang (#!) and nvim should follow suit for the context of that script.

PS. sorry for being all necro

**Comment by Grimy** [comment]
> What are these scripts that everyone keeps talking about?

Plugins. Many vim plugin that use `:!` internally completely break down when `&shell` is not a POSIX shell.

**Comment by wrigleyster** [comment]
sounds to me like those plugins aught to be fixed to use `:!sh -c` instead. `&shell` is for interactive use, it makes no sense to use it as the interpreter of scripts

**Comment by ZyX-I** [comment]
I do not like the complexity of some solutions. My opinion is that:

1. `&shell` and friends must be used each time NeoVim wants anything launched in shell.
2. `&shell` defaults to `$SHELL` on \*nix, but this is the only place where `$SHELL` is used. Defaults are OS-dependent.
3. Script writers are supposed to run `termopen([cmd, arg…])`, `system([cmd, arg…])`, `jobstart([cmd, arg…])` (needs a clear alternative to `:!`). This should be documented in the coding standard and enforced on every runtime file in core.
4. It is supposed that for each function that takes an argv list `func(shellescape(cmd).' '.shellescape(arg)…)` should do nearly the same thing as using argv list. This is neither checked nor enforced though.
5. It is supposed that for each command `exe 'nvimcmd' shellescape(cmd, 1) shellescape(arg, 1)…` should do nearly the same thing as the equivalent function with argv list if any. Again neither checked nor enforced.
6. If we ever have a VimL linter, using `func(string)` and `nvimcmd string`/`exe 'nvimcmd' string` is a warning, `string` containing any character not wrapped with `shellescape()` and not matching `[[:alnum:]\x20]` is an error.
7. `&shell` is *never* assumed to be a POSIX shell. It is also *not* assumed that there *is* a POSIX shell somewhere available. Only assumptions made are the behaviours of `shellescape()` (and `:S` modifier) (with proper other &shell options) and space and ASCII alphanumeric characters outside of `shellescape()`. Reason: Windows users may have no POSIX shell at all on their systems.

Note: `&shell` is *indeed* for interactive use. But I cannot say I never used `system()` in the interactive session, not to mention `!`, `read !` or `write !`.

@wrigleyster You must not assume `sh` is available or is in PATH. There is also Windows. If script writers need scripts to be portable across systems assuming used commands are available and do what expected they write it in VimL using 3. If they target \*nix and need specifically POSIX shell scripts they write a separate file with shebang `#!/bin/sh` and executable bit in VCS. Lengthy code inside a string in VimL without normal line wrapping or multiline string capabilities looks ugly.

---

And let me highlight again: what is the equivalent for `!` that uses argv list? `jobstart` with functions that call `echon` in on_stdout/err callbacks + some message in on_exit callback?

**Comment by wrigleyster** [comment]
@ZyX-I you misunderstand me. My point was: It is the responsibility of the __scriptwriter__ to ensure that commands are interpreted correctly, not the responsibility of vim. Said in other words: if a script breaks because `&shell` has an unexpected value, that is a bug in the script, not in vim.

I am not sure I understand all of your points, but I think we are of the same mind.

**Comment by ZyX-I** [comment]
@wrigleyster I agree that it is script writers responsibility to ensure that commands are interpreted correctly, I only highlighted some reasonable assumptions that may be used if script writer for some reason chooses to use shell in place of argv list. But using `sh -c` explicitly is almost always the wrong way to do this: “those plugins” ought to be fixed by porting their scripts to VimL.

**Comment by masukomi** [comment]
because there still seems to be debate over the correct handling of this ( #4422 )

note that the following quotes are from multiple folks upstream in this discussion

> Many vim plugins that use :! internally completely break down when &shell is not a POSIX shell.

why would you expect a plugin that's dependent upon a particular environment to work outside of that environment? no-one should expect a plugin that requires posix compliance to work in a non-posix environment. The problem here is NOT that the plugin breaks. The problem is that the plugin creator failed to document its requirements OR the installer failed to read them. 

> It rubs me the wrong way that at any point, we might have three different shells launched in three different ways and for different reasons, but I'm not sure this would actually cause any problems. Thoughts?

Maybe i misunderstood your intent, but why would you expect someone who works in shell x all day to have vim open a terminal that resulted in shell y to be happy? All their aliases will be gone, all their functions will be gone, the internal scripting language will be different and usage will degrade dramatically. 


> My point was: It is the responsibility of the script writer to ensure that commands are interpreted correctly, not the responsibility of vim.

yes, 1000 times yes. 

> True, but that begs the question: why not use bash or zsh?

Why not use emacs? The answer is pretty much the same: "Because X does a damn good job and I like the features it provides me" Just because _you're_ happy with the limitations/features of zsh and bash doesn't mean everyone else is.

**Comment by PhilT** [comment]
I hit this problem myself while trying to run external programs on Windows but using GitBash as the terminal. I know this discussion is about what the default should be but here is a workaround for now:

My example uses GitBash (for Windows) as the replacement terminal shell then resets it back to cmd.exe once the terminal opens. It should be possible to tweak for specific use-cases.

```
command! Term :set shell=/Tools/Git/bin/bash.exe|:term
autocmd TermOpen * :set shell=cmd.exe
:cabbrev term <c-r>=(getcmdtype()==':' && getcmdpos()==1 ? 'Term' : 'term')<CR>
```

I hope someone finds this useful.

**Comment by lewis6991** [comment]
This issue hasn't seen real activity in nearly a decade and the overall sentiment seems to be summarised with:

> For a script launching a specific command, rather than relying on `shellescape()` wouldn't it be simpler to just launch the process using the argv-list like with `jobstart` ? (at least when globbing, shell substitution etc is not required) 

and

> My point was: It is the responsibility of the script writer to ensure that commands are interpreted correctly, not the responsibility of vim.

Closing.
